frame/pablo/src/lib.rs

```diff
@@ -467,6 +499,139 @@ pub mod pallet {
+					let rate_base = Self::do_get_exchange_rate(pool_id, false);
+					let rate_quote = Self::do_get_exchange_rate(pool_id, true);
+					match (rate_base, rate_quote) {
+						(Ok(rate_base), Ok(rate_quote)) => PriceCumulativesState::<T>::try_mutate(
+							pool_id,
+							|price_cumulatives| -> Result<Weight, DispatchError> {
+								match *price_cumulatives {
+									// Initialization
+									None => {
+										let (base_price_cumulative, quote_price_cumulative) = (
+											compute_initial_price_cumulative::<T::Convert, _>(
+												rate_base,
+											)?,
+											compute_initial_price_cumulative::<T::Convert, _>(
+												rate_quote,
+											)?,
+										);
+										*price_cumulatives = Some(PriceCumulatives {
+											timestamp: current_timestamp,
+											base_price_cumulative,
+											quote_price_cumulative,
+										});
+										TWAPState::<T>::insert(
+											pool_id,
+											TimeWeightedAveragePrice {
+												base_twap: rate_base,
+												quote_twap: rate_quote,
+											},
+										);
+										Ok(weight + 1)
+									},
+									// Update
+									Some(PriceCumulatives {
+										timestamp: previous_timestamp,
+										base_price_cumulative: previous_price_cumulative_base,
+										quote_price_cumulative: previous_price_cumulative_quote,
+									}) => {
+										let f = |previous_cumulative_price, exchange_rate| {
+											compute_next_price_cumulative::<T::Convert, _, _>(
+												previous_timestamp,
+												previous_cumulative_price,
+												current_timestamp,
+												exchange_rate,
+											)
+										};
+										let (
+											(elapsed, base_price_cumulative),
+											(_, quote_price_cumulative),
+										) = (
+											f(previous_price_cumulative_base, rate_base)?,
+											f(previous_price_cumulative_quote, rate_quote)?,
+										);
+										if elapsed >= T::TWAPInterval::get() {
+											*price_cumulatives = Some(PriceCumulatives {
+												timestamp: current_timestamp,
+												base_price_cumulative,
+												quote_price_cumulative,
+											});
+											let avg = |current: T::Balance,
+											           previous: T::Balance|
+											 -> Result<Rate, DispatchError> {
+												compute_price_average::<T::Convert, _, _>(
+													current, previous, elapsed,
+												)
+											};
+											let average_price_base = avg(
+												base_price_cumulative,
+												previous_price_cumulative_base,
+											)?;
+											let average_price_quote = avg(
+												quote_price_cumulative,
+												previous_price_cumulative_quote,
+											)?;
+											TWAPState::<T>::insert(
+												pool_id,
+												TimeWeightedAveragePrice {
+													base_twap: average_price_base,
+													quote_twap: average_price_quote,
+												},
+											);
+											Ok(weight + 1)
+										} else {
+											Ok(weight)
```

Comment:
This weight calculation seems off.

Reply:
@KaiserKarel  could you please elaborate ? I don't understand what is the issue.